File: 0282.py
INF = 2000000000

class seg():
  a = INF
  b = INF
  def __init__(self, a, b):
    self.a = a
    self.b = b

def seg_min(a, b):
  if(a.a == b.a):
    return b if b.b < a.b else a
  else:
    return b if b.a < a.a else a

def update(i, x):
  i += n - 1
  dat[i] = x
  while(i > 0):
    i = (i - 1) / 2
    dat[i] = seg_min(dat[i * 2 + 1], dat[i * 2 + 2])

def query(a, b, k, l, r):
  if a >= r or b <= l:
    return seg(INF, INF)
  if a <= l and b >= r:
    return dat[k]
  vl = query(a, b, k, l, (l + r) / 2)
  vr = query(a, b, k, (l + r) / 2, r)
  return seg_min(vl, vr)

dat = [seg(INF, INF)] * (1 << 18 + 1)
ans = [0] * (1 << 17)
print len(ans)
print len(dat)

N, R, L = map(int, raw_input().split())
n = 1
while(n < N):
  n *= 2
n *= 2
for i in range(n * 2 - 1):
  dat[i] = seg(INF, INF)
for i in range(N):
  update(i, seg(0, i))

t_old = 0
for i in range(R):
  d, t, x = map(int, raw_input().split())
  d -= 1
  for j in range(n):
    print dat[j].a,
  ans[query(0, N, 0, 0, n).b] += t - t_old
  s = dat[d + n - 1]
  update(d, seg(s.a + x, s.b))

ans[query(0, N, 0, 0, n)] += L - old_t

m = 0
for i in range(N):
  m = max(m, ans[i])
for i in range(N):
  if m == ans[i]:
    print i + 1
    break

File: 0282.cpp
#include<iostream>
#include<map>
#include<algorithm>
using namespace std;

#define INF 2000000000
typedef pair<int, int> seg;
seg dat[1 << 18 + 1];
int ans[1 << 17];

int n, N, R, L;


void update(int i, seg x)
{
  i += n - 1;
  dat[i] = x;
  while(i > 0)
  {
    i = (i - 1) / 2;
    dat[i] = min(dat[i * 2 + 1], dat[i * 2 + 2]);
  }
}

void init()
{
  n = 1;
  while(n < N) n *= 2;
  n *= 2;
  for(int i = 0; i < n * 2 - 1; i++)
    dat[i] = make_pair(INF, INF);
  for(int i = 0; i < N; i++)
    update(i, make_pair(0, i));
}

seg query(int a, int b, int k, int l, int r)
{
  if(a >= r || b <= l)
    return make_pair(INF, INF);
  if(a <= l && b >= r)
    return dat[k];
  seg vr = query(a, b, k, (l + r) / 2, r);
  seg vl = query(a, b, k, l, (l + r) / 2);
  return min(vr, vl);
}


void output()
{
  int m = 0;
  for(int i = 0; i < N; i++)
  {
//    cout << ans[i] << " ";
    m = max(m, ans[i]);
  }
  for(int i = 0; i < N; i++)
  {
    if(ans[i] == m)
    {
      cout << i + 1 << endl;
      return;
    }
  }
}


int main()
{
  int old_t = 0;

  cin >> N >> R >> L;
  init();
  for(int i = 0; i < R; i++)
  {
    int d, t, x;
    cin >> d >> t >> x;
    d--;
    ans[query(0, N, 0, 0, n).second] += t - old_t;
    
    seg s = dat[d + n - 1];
    update(d, make_pair(s.first - x, s.second));
    old_t = t;
  }
  ans[query(0, N, 0, 0, n).second] += L - old_t;

  output();
}
